W-19704481-add-Agent-Fabric-authoring-content-mk (#504)
mulesoft/docs-code-builder: modules/ROOT/pages/af-project-files.adoc

= Reference: Agent Network Project Files

> Your agent network project contains two primary configuration files: agent-network.yaml and exchange.json.

Reviewer: ```suggestion
Your agent network project contains two primary configuration files: `agent-network.yaml` and `exchange.json`.
```

== Basic Concepts

> === Agent-network.yaml

Reviewer: Define the Structure Configuration in the agent-network.yaml File

The agent network YAML file defines a structured configuration for multi-agent systems, enabling orchestration of AI agents with external services, tools, and inter-agent communication. This format provides a declarative way to define agent capabilities, dependencies, and service integrations.

For example, think of the YAML as your organizational chart for digital labor. Each internal organization (for example, HR, IT, Sales, or Engineering) has a hierarchical structure through which actors collaborate. Jobs are split into tasks and assigned.

Just like with human labor, digital agents require coordination. We call these coordinators "brokers". These are special types of agents in your agent network capable of receiving a user prompt, identifying the goal, determining the necessary steps to be carried out, and delegating each step to the proper agent, even if the agent is part of a different broker.

> === Exchange.json

Reviewer: Define Asset Metadata in the exchange.json File

All agent network projects also have an exchange.json file. If you're familiar with Anypoint Platform, you've probably seen this file before. This file contains asset metadata that will be available in Anypoint Exchange after you publish your agent network assets.

== File Elements

=== Exchange.json Elements

> Here's an example file. These key-value pairs are important for your agent network configuration.

Reviewer: ```suggestion
In this example file, these key-value pairs define your agent network configuration.
```

When the agent network is published, all of the assets derived from it share the same groupId (business group) defined in exchange.json.

[source,json]
----
{
  "main": "agent-network.yaml",
  "name": "Employee Onboarding Network",
  "classifier": "agent-network",
  "organizationId": "85de5a54-1f33-4ea4-a1bf-8a65bc409179",
  "descriptorVersion": "1.0.0",
  "tags": [],
  "groupId": "85de5a54-1f33-4ea4-a1bf-8a65bc409179",
  "assetId": "employee-onboarding-network",
  "version": "1.0.5",
  "dependencies": [
    {
      "groupId": "85de5a54-1f33-4ea4-a1bf-8a65bc409179",
      "assetId": "workday-agent",
      "version": "1.0.21",
      "classifier": "agent-metadata",
      "packaging": "zip"
    }
  ],
  "metadata": {
    "variables": {
      "openai": {
        "clientId": {
          "description": "OpenAI LLM Client ID",
          "default": ""
        },
        "clientSecret": {
          "description": "OpenAI LLM Client Secret",
          "default": ""
        },
        "url": {
          "description": "OpenAI URL",
          "default": "",
          "secret": false
        }
      }
    }
  }
}
----

==== Relationship to the Ref Element in Agent-Network.yaml

> The "ref" element is used in different elements in the agent network YAML, and is important for your agent network configuration.

Reviewer: ```suggestion
The "ref" element is used in different elements in the agent network YAML so it's required in your agent network configuration.
```

The "ref" element represents another asset that exists in Exchange. The type of asset referenced depends on the context (agent, MCP server, LLM provider, etc.). The semantic is always that of referencing an asset that exists in Exchange, or will exist when the agent network is published (meaning, it's defined in the file). If referenced, it's in the dependencies section of exchange.json.

> Let's see an example:

Reviewer: ===== Ref Element Example

[source,yaml]
----
     links:
        - agent:
            ref:
              name: workday-agent
----

The above snippet represents a link to a workday agent. The *name* attribute doesn't point to the agent's human friendly or logical name (which can change); instead, it references the *assetId* that workday-agent has in Exchange.

The workday-agent asset can either exist already, or can be defined in this same yaml. In either case, the Exchange *assetId* is the key to this reference.

> In some cases, we'll need the ref element to point to an asset that's defined in a different business group. For this, the *ref#namespace* attribute can be used to reference the asset's groupId.

Reviewer: ```suggestion
In some cases, the ref element needs to point to an asset that's defined in a different business group. For this, the *ref#namespace* attribute can be used to reference the asset's groupId.
```

  maxNumberOfLoops: 25            # Max task steps (default: 25)
  maxConsecutiveErrors: 3         # Max recovery attempts 
  tools: [...]                    # Available tools, including MCP servers
  links: [...]                    # Connected agents
----

===== Llm

The value of this section is a reference to one of the LLMs defined in Anypoint Exchange or in the `llmProviders` section of `agent-network.yaml`. Because it's a reference, you can choose to share the same LLM across all the brokers in your agent network. Or, you can have different brokers use different LLMs to better suit their tasks.

When specifying an LLM model, it must support structured responses in JSON format. For example, a `gpt4.1` model is supported; a `gpt4` model isn't.

[source,yaml]
----
  llm:                        # LLM provider reference
    ref:
name: my-openAI               # Exchange asset ID of this LLM provider. By default, the groupId for this asset is the same as that in exchange.json.
namespace:                    # Optional. If the asset was defined in a different business group, set the corresponding groupId here. 
      configuration:
  model: gpt-5-mini
----

> ===== Instructions (optional)

Reviewer: ```suggestion
===== Instructions
```

Reviewer: I'd not add the "Optional" in the title if it's commented in the description.

        - |
          You're an Incident Management Broker. Your primary responsibility is to coordinate the resolution of incidents reported by customers.
          The process for incident management is:
            1. Fetch Salesforce case details: Retrieve the latest critical case details for the customer.
            2. Fetch entitlement details: Obtain the customer's entitlement information.
            3. Fetch on-call engineer: Identify the current on-call engineer for the incident.
            4. Create Slack war room and invite on-call engineer: Set up a Slack war room channel and invite the on-call engineer.
            5. Summarize actions: Provide a clear, human-readable summary of the steps performed, including information about the created Slack channel and the on-call engineer assigned.
----

You don't need to provide instructions like "split the prompt into tasks", or "select the best tool". The broker does that on its own. These instructions are only necessary for describing specific business processes.

While the instructions section is optional, it improves broker outcomes if you include as much detail about your business process as possible.

===== Links

> This section is the most important part of the entire specification. The links section defines the agents that are available to this broker. Essentially, the links section is where the agent network is defined. It enables interagent communication and orchestration. The broker relies on the agents linked here to execute the appropriate actions to complete a user's goal.

Reviewer: ```suggestion
The links section defines the agents that are available to this broker and it's where the agent network is defined. It enables interagent communication and orchestration. The broker relies on the agents linked here to execute the appropriate actions to complete a user's goal.
```

[source,yaml]
----
links:
  - agent:
      ref:
name: workday-agent	 # Exchange asset ID of the linked agent. By default, the groupId for this asset is the same as that in exchange.json.
namespace:      		 # Optional. If the asset was defined in a different business group, set the corresponding groupId here.
- agent:
ref:
name: badging-agent
- agent:
ref:
name: salesforce-agent
- agent:
ref:
name: zendesk-agent
- agent:
ref:
name: it-agent
----

Keep this section short. Don't add all available agents. Just like `tools`, excess values here leads to drops in agent network accuracy and determinism.

Instead, we recommend a multi-level hierarchical approach. This favors traceability and management and helps keep context sizes in check.

**Links Example**

In this example, the Onboarding Broker orchestrates the process of onboarding a new hire. It orchestrates work with the IT, HR and security brokers, which also happen to be brokers in the same agent network.

> Assets, IT Stock, Workday, Payroll and Badges agents are what we call leaf agents. These aren't brokers but preexisting agents whose purpose isn't to coordinate work, but to act.

Reviewer: ```suggestion
Assets, IT Stock, Workday, Payroll, and Badges agents are what we call leaf agents. These aren't brokers but preexisting agents whose purpose isn't to coordinate work, but to act.
```

image::af-onboarding-broker-example.png["Diagram showing an employee onboarding broker and its agents"]

===== Tools

Tools provide agents with external capabilities. When a broker needs to access an external service that's not another agent, it reaches out to an MCP (Model Context Protocol) server.

[source,yaml]
----
tools:
  - mcp:
ref: 
name: talent-pool-mcp	        # Exchange asset ID of MCP server. By default, the groupId for this asset is the same as that in exchange.json.
allowed:                      # Allowlist specific tools
- TalentPoolMcpServer.match_email_to_address
- mcp:
ref:
name: slack-mcp
allowed:
- SlackMcpServer.send_status_update
----

> By default, the broker has access to all the tools available in the MCP server. However, the most modern LLMs can only handle around 20 to 25 tools per context before starting to hallucinate. It's good practice to limit the available tools to the minimum needed. You can apply that filtering through the allowed list.

Reviewer: ```suggestion
By default, the broker has access to all the tools available on the MCP server. However, the most modern LLMs can only handle around 20 to 25 tools per context before starting to hallucinate. Limit the available tools to the minimum needed. You can apply that filtering through the allowed list.
```

===== Policies

Policies contains the list of policies to apply to the broker. It contains a reference to the policy that needs to be added as a dependency to the agent network project.

Policies can be applied at the broker or at the connection. Reference the policy definition and apply the configuration that wants to be set.

[NOTE]
====
When you add governance policies to your project, those policies persist between deployments. However, if you add governance policies at run-time, those policies won't persist.
====

[source,yaml]
----
policies:
  - ref: 
name:          		# Exchange asset ID of the policy. By default, the groupId for this asset is the same as that in exchange.json.
namespace:     		# Optional. If the policy was defined in a different business group, set the corresponding groupId here.
configuration:
policyConfig1: policyConfig1Value	
----

===== Error Handling

Error scenarios can be handled through these attributes.

[cols="1,3,1,1"]
|===
| Name | Description | Type | Default value

> | maxNumberOfLoops | The maximum number of steps that a task can take. Useful for keeping orchestrations from running too long and consuming too many tokens | Integer | 25

Reviewer: ```suggestion
| maxNumberOfLoops | The maximum number of steps that a task can take. Useful for keeping orchestrations from running too long and consuming too many tokens. | Integer | 25
```

> | maxConsecutiveErrors | The maximum number of errors that the broker will attempt to recover from before returning a failed status. | Integer | 3

Reviewer: ```suggestion
| maxConsecutiveErrors | The maximum number of errors that the broker attempts to recover from before returning a failed status. | Integer | 3
```

|===

==== Agents

Reference agents defined in a different agent network, or another agent used in your company. Any agent is valid, as long as it supports A2A as its communication protocol.

[source,yaml]
----
agents:
workday-agent:			# Used as the assetId of the asset published to Anypoint Exchange. The default groupId is the associated value used in exchange.json.
metadata:
       platform: azure
       protocol: a2a
  badging-agent:
    metadata:
      platform: azure
      protocol: a2a
  salesforce-agent:
    metadata:
      platform: azure
      protocol: a2a
  zendesk-agent:
    metadata:
      platform: azure
      protocol: a2a
  it-agent:
    metadata:
      platform: azure
      protocol: a2a
----

> ==== McpServers

Reviewer: ```suggestion
==== MCP Servers
```

Reviewer: if "mcpServers" is something else, it may be a good idea to enhance this title. Something like "McpServer Attribute".

Author: mcpServers is used in YAML

      authentication:
        kind: oauth2-client-credentials
        token:
          bodyEncoding: form
          timeout: 300
          url: test.com
        clientId: clientId
        clientSecret: clientSecret

  it-agent:
    kind: agent
    ref:
      name: it-agent
    spec:
      url: https://ping-id-a2a-agent-inkbh1.paga8m.usa-w1.cloudhub.io/pingid/
      authentication:
        kind: oauth2-client-credentials
        token:
          bodyEncoding: form
          timeout: 300
          url: test.com
        clientId: clientId
        clientSecret: clientSecret

  my-openAI:
    kind: llm
    ref:
      name: my-openAI
    spec:
      url: ${}
      configuration:
        apiKey: ${openai.apiKey}

  talent-pool-mcp:
    kind: mcp
    ref:
      name: talent-pool-mcp
    spec:
      url: https://talent-pool-mcp-server-inkbh1.paga8m.usa-w1.cloudhub.io/

  slack-mcp:
    kind: mcp
    ref:
      name: slack-mcp
    spec:
      url: https://slack-mcp-server-inkbh1.paga8m.usa-w1.cloudhub.io/
----

===== Connection Resolution

Connection resolution is done at deployment time. The connection reference can be done explicitly or implicitly.

**Explicit**

> In this example, the connection that will be used is **weather-agent-connection**. At deployment, it validates that a connection with the given name exists in the egress gateway on the specified environment.

Reviewer: ```suggestion
In this example, the connection used is **weather-agent-connection**. At deployment, it validates that a connection with the given name exists in the egress gateway on the specified environment.
```
